src/Toml/BiMap.hs

```diff
@@ -51,245 +56,314 @@ module Toml.BiMap
 
 -- | 'Int' bimap for 'AnyValue'. Usually used with 'int' combinator.
-_Int :: BiMap Int AnyValue
+_Int :: TomlBiMap Int AnyValue
 _Int = _BoundedInteger >>> _Integer
 
 -- | 'Float' bimap for 'AnyValue'. Usually used with 'float' combinator.
-_Float :: BiMap Float AnyValue
+_Float :: TomlBiMap Float AnyValue
 _Float = iso realToFrac realToFrac >>> _Double
 
 -- | Helper bimap for 'Text' and strict 'ByteString'
-_ByteStringText :: BiMap ByteString Text
-_ByteStringText = prism T.encodeUtf8 maybeText
-  where
-    maybeText :: ByteString -> Maybe Text
-    maybeText = either (const Nothing) Just . T.decodeUtf8'
+_ByteStringText :: TomlBiMap ByteString Text
+_ByteStringText = prism T.encodeUtf8
+    (\value -> either (\_ -> leftWrongValue . Text $ tShow value) Right $ T.decodeUtf8' value)
 
 -- | 'ByteString' bimap for 'AnyValue'. Usually used with 'byteString' combinator.
-_ByteString:: BiMap ByteString AnyValue
+_ByteString:: TomlBiMap ByteString AnyValue
 _ByteString = _ByteStringText >>> _Text
 
 -- | Helper bimap for 'Text' and lazy 'ByteString'
-_LByteStringText :: BiMap BL.ByteString Text
-_LByteStringText = prism (TL.encodeUtf8 . TL.fromStrict) maybeText
+_LByteStringText :: TomlBiMap BL.ByteString Text
+_LByteStringText = prism (TL.encodeUtf8 . TL.fromStrict) eitherText
   where
-    maybeText :: BL.ByteString -> Maybe Text
-    maybeText = either (const Nothing) (Just . TL.toStrict) . TL.decodeUtf8'
+    eitherText :: BL.ByteString -> Either TomlBiMapError Text
+    eitherText = either (\err -> Left $ ArbitraryError $ tShow err) (Right . TL.toStrict) . TL.decodeUtf8'
 
 -- | Lazy 'ByteString' bimap for 'AnyValue'. Usually used with 'lazyByteString'
 -- combinator.
-_LByteString:: BiMap BL.ByteString AnyValue
+_LByteString:: TomlBiMap BL.ByteString AnyValue
 _LByteString = _LByteStringText >>> _Text
 
 -- | Takes a bimap of a value and returns a bimap of a list of values and 'Anything'
 -- as an array. Usually used with 'arrayOf' combinator.
-_Array :: BiMap a AnyValue -> BiMap [a] AnyValue
+_Array :: TomlBiMap a AnyValue -> TomlBiMap [a] AnyValue
 _Array elementBimap = BiMap
-    { forward = mapM (forward elementBimap) >=> fmap AnyValue . toMArray
-    , backward = \(AnyValue val) -> matchArray (backward elementBimap) val
+    { forward = mapM (forward elementBimap) >=> fmap AnyValue . toTomlBiMapError . toMArray
+    , backward = \(AnyValue val) -> toTomlBiMapError $ matchArray (toMatchError . backward elementBimap) val
     }
 
 -- | Takes a bimap of a value and returns a bimap of a non-empty list of values
 -- and 'Anything' as an array. Usually used with 'nonEmptyOf' combinator.
-_NonEmpty :: BiMap a AnyValue -> BiMap (NE.NonEmpty a) AnyValue
-_NonEmpty bimap = BiMap (Just . NE.toList) NE.nonEmpty >>> _Array bimap
+_NonEmpty :: TomlBiMap a AnyValue -> TomlBiMap (NE.NonEmpty a) AnyValue
+_NonEmpty elementBimap = BiMap
+    { forward = mapM (forward elementBimap) >=>
+        fmap AnyValue . toTomlBiMapError . toMArray . NE.toList
```

Comment:
I propose to remove some code duplication between this function and `_Array`. You can implement:

```haskell
_NonEmptyArray :: TomlBiMap (NonEmpty a) [a]
```

where you handle error of empty list. And then you can implement:

```haskell
_NonEmpty bimap = _NonEmptyArray >>> _Array bimap
```